arangod/Cluster/ActionBase.cpp

```diff
@@ -0,0 +1,261 @@
+
+} // ActionBase::done
+
+ActionDescription const& ActionBase::describe() const {
+  return _description;
+}
+
+MaintenanceFeature& ActionBase::feature() const {
+  return _feature;
+}
+
+VPackSlice const ActionBase::properties() const {
+  return _description.properties()->slice();
+}
+
+
+/// @brief Initiate a new action that will start immediately, pausing this action
+void ActionBase::createPreAction(std::shared_ptr<ActionDescription> const & description) {
+
+  _preAction = description;
+  _feature.preAction(description);
+
+  // shift from EXECUTING to WAITINGPRE ... EXECUTING is set to block other
+  //  workers from picking it up
+  if (_preAction) {
+    setState(WAITINGPRE);
+  } else {
+    _result.reset(TRI_ERROR_BAD_PARAMETER, "preAction rejected parameters.");
+  } // else
+
+} // ActionBase::createPreAction
+
+
+/// @brief Retrieve pointer to action that should run before this one
+std::shared_ptr<Action> ActionBase::getPreAction() {
+  return (_preAction != nullptr) ? _feature.findAction(_preAction) : nullptr;
+}
+
+
+/// @brief Retrieve pointer to action that should run after this one
+std::shared_ptr<Action> ActionBase::getPostAction() {
+  return (_postAction != nullptr) ? _feature.findAction(_postAction) : nullptr;
+}
+
+
+/// @brief Create a new action that will start after this action successfully completes
+void ActionBase::createPostAction(std::shared_ptr<ActionDescription> const& description) {
+
+  // preAction() sets up what we need
+  _postAction = description;
+  _feature.postAction(description);
+
+  // shift from EXECUTING to WAITINGPOST ... EXECUTING is set to block other
+  //  workers from picking it up
+  if (_postAction) {
+    setState(WAITINGPOST);
+  } else {
+    _result.reset(TRI_ERROR_BAD_PARAMETER, "preAction rejected parameters for _postAction.");
+  } // else
+
+} // ActionBase::createPostAction
+
+
+void ActionBase::startStats() {
+
+// done when first  _actionStarted = std::chrono::system_clock::now();
+
+} // ActionBase::startStats
+
+
+void ActionBase::incStats() {
+
+  ++_progress;
+  //_actionLastStat = std::chrono::system_clock::now();
+
+} // ActionBase::incStats
+
+
+void ActionBase::endStats() {
+
+  //_actionDone = std::chrono::system_clock::now();
+
+} // ActionBase::endStats
+
+
+Result arangodb::actionError(int errorCode, std::string const& errorMessage) {
+  LOG_TOPIC(ERR, Logger::MAINTENANCE) << errorMessage;
+  return Result(errorCode, errorMessage);
+}
+
+Result arangodb::actionWarn(int errorCode, std::string const& errorMessage) {
+  LOG_TOPIC(WARN, Logger::MAINTENANCE) << errorMessage;
+  return Result(errorCode, errorMessage);
+}
+
+void ActionBase::toVelocyPack(VPackBuilder & builder) const {
+  VPackObjectBuilder ob(&builder);
+
+  builder.add("id", VPackValue(_id));
+  builder.add("state", VPackValue(_state));
+  builder.add("progress", VPackValue(_progress));
+#if 0  /// hmm, several should be reported as duration instead of time_point
```

Comment:
I set this to #if 0 because I was too lazy to figure out the threading issues.  You fixed threading with the atomic uint64_t stuff.  Should reactivate this code using your timepointToString() for all four.  Ignore my duration versus time_point comment (just erase it).

Reply:
fixed